### DISCUSSION TITLE
Question about the PETSc options

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#22344] [Q&A General] [answered]
I used a strange method to write a PDE which contains the variables of different blocks in Kernel. 
But it shows `Linear solve did not converge due to DIVERGED_BREAKDOWN` in terminal. I'm trying to figure out where the problem is.
I guess maybe the method is wrong or the form of the PDE is wrong.
I noticed there is a PETSc options `-pc_type svd -pc_svd_monitor` in NonlinearSystemBase[](https://mooseframework.inl.gov/source/systems/NonlinearSystemBase.html). And it says _"You can inspect the condition number of your preconditioning matrix for small problems (should be less than 1000 degrees of freedom) by running it"_. I'm wondering if this option can show the form of the PDE is true or false. I want to know if this can do what I want and how to use this PETSc option?

**Comment by GiudGiud** [comment, marked as answer]
Hello

You pass these parameters to petsc_options_iname and petsc_options_ivalue in the Executioner block.

This wont tell you if the PDE is written properly or not, but it will `roughly` tell you if it's solvable. If the conditioning is decent, PETSc has generally no trouble solving the linear problems. 

If you want to test if you implemented the PDE right you need to perform a verification procedure. 
I ll paste here @hugary1995 's excellent suggestions from another post:
```
 In FEM code, typically two methods are used for the verification stage in code development

Patch test: You'll need to setup a problem that is as simple as possible in which the analytical solution is known, and the solution can
be exactly matched by the basis function of your choice. For example, you can create a single-element cube, apply Dirichlet BC on
 one side, and apply Neumann BC on the opposite side. By applying your kernels one at a time, you can easily match the numerical
 solution (given by MOOSE) with the analytical solution.

Method of manufactured solution: This method is documented here https://mooseframework.inl.gov/python/mms.html. In essence, you
 will first manufacture an arbitrary, non-homogeneous solution, and then work your way back to derive the corresponding "body force"
 that would result in the manufactured solution. Even if the manufactured solution can't be exactly represented on the basis function
 of your choice, you should still see optimal order of convergence by looking at the appropriate error measure.
```

Another concern is whether you implemented the jacobian right (did you use AD? If so dont worry about that part).
This page has resources for checking this:
https://mooseframework.inl.gov/help/development/analyze_jacobian.html

**Reply by Ethan-xj** [reply]
Yes I've used AD. Thank you!
I think I should test if I implemented the PDE right first.
